File: Problem Solving/Algorithms/Implementation/Non-Divisible Subset.py
#!/bin/python3

import math
import os
import random
import re
import sys

#
# Complete the 'nonDivisibleSubset' function below.
#
# The function is expected to return an INTEGER.
# The function accepts following parameters:
#  1. INTEGER k
#  2. INTEGER_ARRAY s
#

def nonDivisibleSubset(k, s):
    # Write your code here
    remainder_count = [0] * k
    for item in s:
        remainder_count[item % k] += 1

    count = min(remainder_count[0], 1)
    for i in range(1, (k // 2) + 1):
        if i == k - i:
            count += min(remainder_count[i], 1)
        else:
            count += max(remainder_count[i], remainder_count[k - i])

    return count


if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    first_multiple_input = input().rstrip().split()

    n = int(first_multiple_input[0])

    k = int(first_multiple_input[1])

    s = list(map(int, input().rstrip().split()))

    result = nonDivisibleSubset(k, s)

    fptr.write(str(result) + '\n')

    fptr.close()

File: Problem Solving/Algorithms/Implementation/Non-Divisible Subset.cpp
#include <bits/stdc++.h>

using namespace std;

string ltrim(const string &);
string rtrim(const string &);
vector<string> split(const string &);

/*
 * Complete the 'nonDivisibleSubset' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts following parameters:
 *  1. INTEGER k
 *  2. INTEGER_ARRAY s
 */

    // remainder_count = [0] * k
    // for item in s:
    //     remainder_count[item % k] += 1

    // count = min(remainder_count[0], 1)
    // for i in range(1, (k // 2) + 1):
    //     if i == k - i:
    //         count += min(remainder_count[i], 1)
    //     else:
    //         count += max(remainder_count[i], remainder_count[k - i])

    // return count


int nonDivisibleSubset(int k, vector<int> s) {
    vector<int> remainders;
    for (int i = 0; i < k; i++)
        remainders.push_back(0);

    for (auto item: s)
        remainders[item % k]++;

    int count = min(remainders[0], 1);
    for (int i = 1; i <= k / 2; i++)
        if (i == k - i)
            count += min(remainders[i], 1);
        else
            count += max(remainders[i], remainders[k - i]);

    return count;
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string first_multiple_input_temp;
    getline(cin, first_multiple_input_temp);

    vector<string> first_multiple_input = split(rtrim(first_multiple_input_temp));

    int n = stoi(first_multiple_input[0]);

    int k = stoi(first_multiple_input[1]);

    string s_temp_temp;
    getline(cin, s_temp_temp);

    vector<string> s_temp = split(rtrim(s_temp_temp));

    vector<int> s(n);

    for (int i = 0; i < n; i++) {
        int s_item = stoi(s_temp[i]);

        s[i] = s_item;
    }

    int result = nonDivisibleSubset(k, s);

    fout << result << "\n";

    fout.close();

    return 0;
}

string ltrim(const string &str) {
    string s(str);

    s.erase(
        s.begin(),
        find_if(s.begin(), s.end(), not1(ptr_fun<int, int>(isspace)))
    );

    return s;
}

string rtrim(const string &str) {
    string s(str);

    s.erase(
        find_if(s.rbegin(), s.rend(), not1(ptr_fun<int, int>(isspace))).base(),
        s.end()
    );

    return s;
}

vector<string> split(const string &str) {
    vector<string> tokens;

    string::size_type start = 0;
    string::size_type end = 0;

    while ((end = str.find(" ", start)) != string::npos) {
        tokens.push_back(str.substr(start, end - start));

        start = end + 1;
    }

    tokens.push_back(str.substr(start));

    return tokens;
}
